Kernel/Devices/Storage/VirtIO/VirtIOBlockDevice.cpp

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2023, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Devices/Storage/VirtIO/VirtIOBlockDevice.h>
+#include <Kernel/Memory/MemoryManager.h>
+#include <Kernel/Tasks/WorkQueue.h>
+
+namespace Kernel {
+
+namespace VirtIO {
+
+static constexpr u64 VIRTIO_BLK_F_BARRIER = 1ull << 0;       // Device supports request barriers.
+static constexpr u64 VIRTIO_BLK_F_SIZE_MAX = 1ull << 1;      // Maximum size of any single segment is in size_max.
+static constexpr u64 VIRTIO_BLK_F_SEG_MAX = 1ull << 2;       // Maximum number of segments in a request is in seg_max.
+static constexpr u64 VIRTIO_BLK_F_GEOMETRY = 1ull << 4;      // Disk-style geometry specified in geometry.
+static constexpr u64 VIRTIO_BLK_F_RO = 1ull << 5;            // Device is read-only.
+static constexpr u64 VIRTIO_BLK_F_BLK_SIZE = 1ull << 6;      // Block size of disk is in blk_size.
+static constexpr u64 VIRTIO_BLK_F_SCSI = 1ull << 7;          // Device supports scsi packet commands.
+static constexpr u64 VIRTIO_BLK_F_FLUSH = 1ull << 9;         // Cache flush command support.
+static constexpr u64 VIRTIO_BLK_F_TOPOLOGY = 1ull << 10;     // Device exports information on optimal I/O alignment.
+static constexpr u64 VIRTIO_BLK_F_CONFIG_WCE = 1ull << 11;   // Device can toggle its cache between writeback and writethrough modes.
+static constexpr u64 VIRTIO_BLK_F_DISCARD = 1ull << 13;      // Device can support discard command, maximum discard sectors size in max_discard_sectors and maximum discard segment number in max_discard_seg.
+static constexpr u64 VIRTIO_BLK_F_WRITE_ZEROES = 1ull << 14; // Device can support write zeroes command, maximum write zeroes sectors size in max_write_zeroes_sectors and maximum write zeroes segment number in max_write_zeroes_seg.
+
+static constexpr u64 VIRTIO_BLK_T_IN = 0;
+static constexpr u64 VIRTIO_BLK_T_OUT = 1;
+static constexpr u64 VIRTIO_BLK_T_FLUSH = 4;
+static constexpr u64 VIRTIO_BLK_T_GET_ID = 8;
+static constexpr u64 VIRTIO_BLK_T_GET_LIFETIME = 10;
+static constexpr u64 VIRTIO_BLK_T_DISCARD = 11;
+static constexpr u64 VIRTIO_BLK_T_WRITE_ZEROES = 13;
+static constexpr u64 VIRTIO_BLK_T_SECURE_ERASE = 14;
+
+static constexpr u64 VIRTIO_BLK_S_OK = 0;
+static constexpr u64 VIRTIO_BLK_S_IOERR = 1;
+static constexpr u64 VIRTIO_BLK_S_UNSUPP = 2;
+
+struct [[gnu::packed]] virtio_blk_config {
```

Comment:
These definitions look very similar to those from the Linux UAPI. 

https://elixir.bootlin.com/linux/v6.9/source/include/uapi/linux/virtio_blk.h
https://github.com/qemu/qemu/blob/9360070196789cc8b9404b2efaf319384e64b107/include/standard-headers/linux/virtio_blk.h

Which is fine because the header explicit says it's licensed as BSD 3-clause so that other projects can implement drivers, but those headers are third-party and need their license intact if we are going to goink them.

Can you:
1. Move these definitions into a `Defintions.h` with the license comment header reproduced in full
2. Add an exception for that file in our check-license-headers linter script
3. Change the struct names to follow our style? (i.e. PascalCase rather than snake_case)

Reply:
Those definitions look like they probably are just from the virtio spec.
I'm not sure under what license the spec code snippets are though.

Reply:
Thanks for the review!

The definitions are in fact not copied from any other project. I didn't look at the Linux kernel implementation at all and did not copy QEMU definitions (although did look at them).

The definitions come directly from the spec here:
https://docs.oasis-open.org/virtio/virtio/v1.2/csd01/virtio-v1.2-csd01.html#x1-2740002
Added a reference to the spec.

I am not 100% sure that we don't need to add the copyright notice of the spec authors (belonging to OASIS Open). But neither QEMU not Linux kernel reference OASIS copyright. So I guess, that's fine?..

Now about the names. I strived to keep modifications to the definitions from the spec to the minimum. I felt like it would be unnecessarily confusing if we change the naming style compared to the spec. Note, that the "non-compliant names" are localised to the VirtIO namespace.

However vaguely remember having the same discussion in https://github.com/SerenityOS/serenity/pull/19757 and I ended up updating to PascalCase. So did the same here.